apps/rabbitmq_prelaunch/src/rabbit_boot_state_systemd.erl

```diff
@@ -0,0 +1,160 @@
+%%%-------------------------------------------------------------------
+%% The contents of this file are subject to the Mozilla Public License
+%% Version 1.1 (the "License"); you may not use this file except in
+%% compliance with the License. You may obtain a copy of the License
+%% at https://www.mozilla.org/MPL/
+%%
+%% Software distributed under the License is distributed on an "AS IS"
+%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
+%% the License for the specific language governing rights and
+%% limitations under the License.
+%%
+%% The Original Code is RabbitMQ.
+%%
+%% The Initial Developer of the Original Code is GoPivotal, Inc.
+%% Copyright (c) 2007-2020 Pivotal Software, Inc.  All rights reserved.
+%%
+
+-module(rabbit_boot_state_systemd).
+
+-behavior(gen_server).
+
+-export([start_link/0]).
+
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-record(state, {sd_notify_module,
+                socket}).
+
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init([]) ->
+    case code:load_file(sd_notify) of
+        {module, sd_notify} ->
+            {ok, #state{sd_notify_module = sd_notify}};
+        {error, _} ->
+            case rabbit_prelaunch:get_context() of
```

Comment:
The `rabbit_boot_state_sup` tree is started before `rabbit_prelaunch` runs, so there is a chance this process also runs before the context is initialized. Now that we have that, we should perhaps revert part of one of my previous commit: this module should check `$NOTIFY_SOCKET` itself to make it independent (and RabbitMQ itself doesn't care about this variable after the systemd support is properly isolated). The revert can be handled later.

Reply:
This crossed my mind, but in practice the context seemed to have been initialized. Another option that I considered was to start cleanly in init, and gen_server:call another init message to the listeners in prelaunch. This could occur after prelaunch logging is set up, and the boot state listeners could stop cleanly then.

Reply:
I think we should just read `$NOTIFY_SOCKET` value and avoid any race conditions. This is not a RabbitMQ configurable variable anyway. I should not have added it to `rabbit_env`.

Reply:
Do you agree to read `$NOTIFY_SOCKET` directly in this function, instead of relying on the context?

Reply:
Yes, I agree. I'll make the change.

Reply:
Turns out the context was not yet properly initialized, so the change has been made.